Report builder

JsonHrdwick opened the pull request:
Beginning commits for the report builder function. Not functional currently but wanted to get your feedback on progress during our next meeting.

bowring commented:
good start, but somehow I have miscommunicated what is needed:

the report builder is independent of data, but  rather organizes the user functions
each specified report structure can be saved, edited, deleted, etc.
see screenshot of ET_Redux report settings manager - it provides a way to organize reports without showing any data
see screenshot of Squid report settings manager - ditto
Et_Redux is rudimentary, Squid uses drag n drop, which I prefer.

maybe we need a quick meeting on Monday?


![Untitled 2](https://github.com/user-attachments/assets/293dfce1-a50e-45e4-9ffb-0019c0c61c7b)

![Untitled 33](https://github.com/user-attachments/assets/a1a88066-3a33-4421-9d4f-55073fe5c791)

JsonHrdwick commented:
I'm available Monday. I understand now what you are looking for. I think most of my confusion came from redux and the other existing report functions. If you want to nail down how the ui is laid out in the meeting that'd make designing it a lot easier.

bowring commented:
Name your time😊

Sent from Gmail Mobile


On Sun, Mar 2, 2025 at 11:06 AM Jason Hardwick ***@***.***>
wrote:

> I'm available Monday. I understand now what you are looking for. I think
> most of my confusion came from redux and the other existing report
> functions. If you want to nail down how the ui is laid out in the meeting
> that'd make designing it a lot easier.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CIRDLES/Tripoli/pull/291#issuecomment-2692796699>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIK2Y3G3UR4Q23QLMLF5ZD2SMUA3AVCNFSM6AAAAABXJ3MAQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJSG44TMNRZHE>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
> [image: JsonHrdwick]*JsonHrdwick* left a comment (CIRDLES/Tripoli#291)
> <https://github.com/CIRDLES/Tripoli/pull/291#issuecomment-2692796699>
>
> I'm available Monday. I understand now what you are looking for. I think
> most of my confusion came from redux and the other existing report
> functions. If you want to nail down how the ui is laid out in the meeting
> that'd make designing it a lot easier.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CIRDLES/Tripoli/pull/291#issuecomment-2692796699>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIK2Y3G3UR4Q23QLMLF5ZD2SMUA3AVCNFSM6AAAAABXJ3MAQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJSG44TMNRZHE>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

JsonHrdwick commented:
Can you do 10am?

JsonHrdwick commented:
Current iteration:
- Redesigned UI as data agnostic
- Report saving/importing works
- Implemented design features from Squid UI
- Left room in UI for placement of accordion or button menu such as in Squid/Redux
- Matched Tripoli design styles
- Implemented "unsaved changes" check (needs further refinement)
- Save/Cancel(Changes)/Rename(Report)/Delete/Import buttons work
- Defined Test function for report comparison (encountered an issue where reports were return unequal using Objects.equals() for TreeSet<ReportCategory>. Could not replicate issue in test environment but manually iterating through the sets fixed the issue)


Missing:
- Nested class structures are not being de/serialized properly (data is lost)
- No way to add columns to new category currently (No functionality exists)
- No way to view Report Name (Waiting for suggestion)
- CustomReportMenu does not repopulate after initial loading
- Export/Copy/New Buttons do not function
- List views do not update after loading a new report or reverting changes

bowring commented:
looking great!

need a way to turn off/on columns and categories - right mouse click with strike-through ?

JsonHrdwick commented:
I missed that in my summary but it is implemented and tracked! Strike
through and grayed out

On Fri, Mar 7, 2025, 1:10 PM Jim Bowring ***@***.***> wrote:

> looking great!
>
> need a way to turn off/on columns and categories - right mouse click with
> strike-through ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CIRDLES/Tripoli/pull/291#issuecomment-2707089744>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLS4A545QX65UZKJN4FL75D2THOHXAVCNFSM6AAAAABXJ3MAQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGA4DSNZUGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
> [image: bowring]*bowring* left a comment (CIRDLES/Tripoli#291)
> <https://github.com/CIRDLES/Tripoli/pull/291#issuecomment-2707089744>
>
> looking great!
>
> need a way to turn off/on columns and categories - right mouse click with
> strike-through ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CIRDLES/Tripoli/pull/291#issuecomment-2707089744>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLS4A545QX65UZKJN4FL75D2THOHXAVCNFSM6AAAAABXJ3MAQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBXGA4DSNZUGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

JsonHrdwick commented:
Added:
- Implemented report name in the title bar
- Unsaved changes are now properly tracked as well as manual backtracking of changes
- Unsaved changes indicator in the title bar (* after name)
- Main page report menu now reflects updates to reports. Menu list is reloaded when the bar is opened.
- List view correctly repopulates when loading a new report
- Serialization works (Always did, this was a local issue)
- Serialization test made

Missing:
- [ ] Waiting on column adding functionality
- [ ] Waiting on mutli analysis handing
- [ ] Waiting on direction for missing button handling (New/Export/Copy)
- [ ] No way to delete a category once added (Right click menu to delete? Add a delete category button?)

bowring commented:
looks good!
various in no order:
- [ ] need face time soon
- [ ] use hand pointer for columns drag also
- [ ] make report window open centered on app (see About window)
- [ ]  category and column  d&d breaks on dragging to white space at bottom of list
- [ ] need a "peek" view as per squid that shows the value of the selected field for each of the analyses in the session
- [ ] need a write to csv and open file command again per Squid
- [ ] ditto with a quickie view of report
- [ ] need way to specify columns in new category - consider view to left of all columns per squid accordions - listing ratios separate from functions - OR allow drag and drop from columns of one category to another category - NOTE currently drag and drop responds incorrectly when column dragged to category inckuding the error of dropping a column onto the white space of the category listing
- [ ] create category does not check for duplication of name
- [ ] need to be able to delete custom categories
- [ ] implement width and height bindings to make internal panes resize with report window (see plots)

JsonHrdwick commented:
Scheduled meeting for this Wed. Cant remember if you were needing to reschedule that but we can meet anytime this weekend or monday if you need to move it around.

Was waiting on your feedback for some of those missing features.

JsonHrdwick commented:
Current Milestone:
- Hand pointer for column view
- No more cross drop between LVs and fixed index error for drop onto whitespace
- Locked analysis name column except for when pulled from accordion
- Replaced default verbiage and cleaned related code
- Implemented 'new' button and removed unused buttons
- Right click to remove cell implemented
- Report Builder opens center to primary app window and is resizable
- Column adding via accordion working
- Font set to same as rest of project

Missing:
The data peaking and report generation from last meeting

JsonHrdwick commented:
- Fixed bugs mentioned in quick meeting
- column detail peek window
- generate report as CSV (See img)

![image](https://github.com/user-attachments/assets/47d80b01-37e6-4905-a2dc-7f721976682f)

Custom expressions still not implemented yet
Do the category names need to be included in the report? I was prioritizing readability (hence the dual header for UFs) over data processing efficiency
Other QoL improvements?

JsonHrdwick commented:
Identified the bug I spoke about. Issue lied in making the field for methodName into static type (was using it statically in my method call). Reverting it to non-static and fixing my method to also be non-static fixed the issue.

Also pushed a fix for the method name issue in the Analysis.java class which caused the files with .TIMS** types to be truncated. Here is my fix for that:
`            extractedAnalysisName = dataFilePath.toFile().getName().substring(0, dataFilePath.toFile().getName().lastIndexOf('.')); 
`

JsonHrdwick commented:
Changes:
- Method Name display changed to a label
- Full Report now displays (Not Editable) on the name plates for increased clarity
- Removed Rename button and moved functionality to Report Name textbox editing and saving
- New category textbox text is red (70% opacity) to increase readability
- Reworked the TripoliMessageDialogs to be more clear in the Report Builder
- Fixed bug in TripoliMessageDialog where user could not exit out of dialog
- Refactored .trf file extension to .tripreport
- Creating a new report of any type will initialize it with the analysis name as the report name
- Moved location for generated CSV reports to the same directory as the active analysis
- Restructured the column details peek view to show analysis name in first column and selected ReportColumn in 2nd column
- Restructured UF output to follow this format: "{UF name} Mean, StdDev, Variance" where each label is a separate header column
- UF data is rounded to 4 significant figures
- Added file metadata to end of the generated report
- Added custom expression category as well as boolean flag in the UserFunction class

bowring commented:
@JsonHrdwick - good work!
note:
I updated this branch to align with main.

- [ ] tripReport names should be for method, while csv file names should mix session name and analysis name(s)
- [ ] When tripReport is saved, no need to show where, just do it.
- [ ]  On tripReport save confirm if over-writing
- [ ]  On tripReport edit, the asterisk is confusing as it does not survive the save although the dialog shows the asterisk.  Instead add label that says edit mode in red
- [ ] On tripReport delete no need for second confirmation unless error occurred
- [ ] when reports manager first opens, peek does not work properly until a change in selected category or column is made
- [ ] the reports manager window should be a singleton, currently it opens new copies on top of each other
- [ ]  for multiple analyses, need to modify list of analyses to allow multiple selection (flag on list object) and if all selected have the same method, then one report Manager window should open and peek will have a line for each analysis, as will the generated csv report.  If methods don't match, alert user and abort.
- [ ] In reports Manager, in the accordions on left, modify list objects to allow multi-select and drag and drop of multi-select

Cheers, @bowring

JsonHrdwick commented:
Implemented all suggested improvements.
- Session manager can now multi-select analyses and the selected ones will be passed to the report builder. A dialog prevents selecting multiple method names.
- Currently the output location for a generated CSV is the location of the first analysis in the list.
- CSV name is in the style requested, which is a concatenation of the SessionName and the name of each analysis.
- I considered moving the (Not Editable) tag to the same label as UNSAVED CHANGES text since there isnt any overlap between the two but I thought it made the consistency of the label confusing. Let me know what you think.

bowring commented:
@JsonHrdwick - nice work;  I am merging into main;
